### DISCUSSION TITLE
Доработка скрипта

### DISCUSSION THREAD

**Discussion by ScreNN** [Chimildic/goofy#206] [Помощь с алгоритмом] [answered]
Здравствуйте!

Накидал скрипт, который стягивает Дневной плейлист из ЯМы в Спотифай

```
// Получить Плейлист дня из Яндекса
function getYandexDailyPlaylist(){
  let savedTracks = Source.getSavedTracks();
  let playlistTracks = gooex.Playlist.getTracks('id', 'yamusic-daily');
  let strTracks = yandexTracksToString_(playlistTracks);
  let spotifyTracks = Search.multisearchTracks(strTracks);
  let banTracks = Cache.read('Banned.json');
  Combiner.push(banTracks, savedTracks);

// Удалить из выборки лайкнутые треки и треки из бан-листа
  Filter.removeTracks(spotifyTracks, banTracks);
  Filter.matchOriginalOnly(spotifyTracks);

// Передать вывод в Spotify
  Playlist.saveWithReplace({
    id: '5sdCRosGueuzFgctiwbWn8',
    name: '💛 Плейлист дня',
    tracks: spotifyTracks,
    description: 'Плейлист дня из Яндекса',
    public: false,
  });
}

function yandexTracksToString_(tracks) {
  return tracks.map(t => {
    t = t.track || t;
    return `${t.artists[0].name} ${t.title}`.clearName();
  });
}
```
Все работает отлично, но есть ощущение, что можно еще доработать некоторые моменты. Например, Яндекс подбирает треки в плейлист дня иногда достаточно некрасиво, в том плане что треки туда попадают не из оригинальных альбомов или EP, а из подборок, типа "Лучшие треки за 2021" или "Хит парад траливали" и.т.д. В итоге, эти треки попадают в Спотифай и вытягиваются так же из этих подборок, потому что такие подборки есть и в Спотифае. Хотелось бы сделать как то так, чтобы треки брались не из подборок, а из оригинальных альбомов и EP. Этот момент как то обойти можно? Фильтрация `Filter.matchOriginalOnly(spotifyTracks);` в этом плане не спасает. И нужна ли она вообще тут?

Далее, в скрипте есть чтение данных из Бан-файла
`let banTracks = Cache.read('Banned.json');`
Бан-файл формируется из Бан-листа по триггеру раз в час, например. Это нужно, чтобы Спотифай не индексировал Бан-лист, и не предлагал рекомендации на его основе. Есть такая штука, что Спотифай индексирует все плейлисты пользователя, чтобы давать рекомендации. А так как в Бан-лист входят треки которые не нужны, то этот Бан-лист перенесен в Бан-файл.
```
function banTracksCache() {
  let banTracksFile = Source.getPlaylistTracks('', '1l6fDXBOpvthH3Wd39uH0w');
  Cache.append('Banned.json', banTracksFile);
  Playlist.removeTracks('1l6fDXBOpvthH3Wd39uH0w', banTracksFile);
}
```
Последняя строка отвечает за очистку Бан-листа, после отработки функции, чтобы Бан-лист не отсвечивал.
По этому блоку кода есть вопрос. Как правильно сделать проверку на существования файла Banned.json через инструкцию if? То есть, если файл существует, то выполнить условие `Cache.append('Banned.json', banTracksFile);` и дописать данные в файл Banned.json. Если же файла не существует, то выполнить условие `Cache.write('Banned.json', banTracksFile);`, создать файл и записать данные в массив

**Comment by ScreNN** [comment]
Немного модифицировал функцию генерации Бан-файла
```
function banTracksCache() {
  let banTracksFile = Source.getPlaylistTracks('', '1l6fDXBOpvthH3Wd39uH0w');
  Cache.append('BannedTracks.json', banTracksFile);
  Playlist.removeTracks('1l6fDXBOpvthH3Wd39uH0w', banTracksFile);

  let filename = 'BannedTracks.json';
  let tracks = Cache.read(filename);
  Cache.compressTracks(tracks);
  Cache.write(filename, tracks);
}
```
Добавил компрессор. Я думаю, так будет правильнее

**Comment by Chimildic** [comment, marked as answer]
1. До сих пор никто не приходил с задачей подмены микс-альбома на оригинал. Поэтому готового решения нет. Я даже не могу представить по какому формальному признаку это можно определить при импорте. 

   Как видите по функции _yandexTracksToString_, название альбома не участвует. Отсюда и Spotify не особо виноват в том, что находит. Может этот альбом просто популярнее? 

   Другой вариант, название трека содержит явное упоминание версии трека: mix, cover, feat и т.д. В таком случае их можно отрезать. Для этого добавьте функцию `parseNameTrack` и укажите ее для `multisearchTracks`

```js
  // ...
  let spotifyTracks = Search.multisearchTracks(strTracks, parseNameTrack)
  // ...

  function parseNameTrack(item) {
    let re = /( mix| feat| cover|,|\/|[(]|[[])/gi;
    return item.split(re)[0];
  }
```

2. Функции типа _match*_ работают только со строковым названием. По принципу _содержит_ или _является_ таким-то алфавитом. Логики какого-то поиска, замены там нет.

3. Согласно описанию [Cache.append](https://chimildic.github.io/goofy/#/reference/cache?id=append):

    > Присоединить данные к массиву из файла. Создает файл, если его не существует.

    Поэтому блок _if-else_ не нужен.

**Reply by ScreNN** [reply]
Спасибо.
Ну а в общем, код выглядит нормально? Никаких сюрпризов не должно вылезти? :)

Хм.. А как так вышло, что у меня вчера через Cache.append файл сам не создался. Я по этому и уточнил это момент. Странно

**Reply by Chimildic** [reply]
Возможно добавляемый массив был пустым. В таком случае файл не создается.

**Reply by ScreNN** [reply]
Ага, возможно кстати. Понял. Спасибо.
По коду значит все нормально, замечаний нету. Ну, вроде все работает как задумано. Там еще ниже про переменную `let banTracks = [];` можете ответить?

**Reply by Chimildic** [reply]
Пишите в случае возникновения ошибок. Чего-то явно необходимого нет. Оптимизацией заниматься рано.

**Comment by ScreNN** [comment]
И еще момент всплыл в голове.
Если взять стандартный шаблон
```
/**
 * Объединить треки из личных миксов в один плейлист
 * Удалить дубликаты, историю прослушиваний, лайки
 */
function updateUnknownSet() {
    let banTracks = [];
    let savedTracks = Source.getSavedTracks();
    let recentTracks = RecentTracks.get(1000);
    Combiner.push(banTracks, savedTracks, recentTracks);

    // Вставьте id плейлистов, которые берутся из ссылки или URI
    // Пример: https://chimildic.github.io/goofy/#/reference/desc?id=Идентификатор
    let onlyForYouTracks = Source.getTracks([
        { name: 'Микс дня 1', id: 'вашеId' },
        { name: 'Микс дня 2', id: 'вашеId' },
        { name: 'Микс дня 3', id: 'вашеId' },
        { name: 'Микс дня 4', id: 'вашеId' },
        { name: 'Микс дня 5', id: 'вашеId' },
        { name: 'Микс дня 6', id: 'вашеId' },
        { name: 'Радар новинок', id: 'вашеId' },
        { name: 'Открытия недели', id: 'вашеId' },
    ]);
    
    Filter.dedupTracks(onlyForYouTracks);
    Filter.dedupArtists(onlyForYouTracks);
    Filter.removeTracks(onlyForYouTracks, banTracks);
    Filter.matchOriginalOnly(onlyForYouTracks);
    Filter.matchExceptRu(onlyForYouTracks);

    // Пример удаления исполнителей, если ведете бан-плейлист
    // let banArtists = Source.getPlaylistTracks('', 'вашеId');
    // Filter.removeArtists(onlyForYouTracks, banArtists);
    
    Filter.rangeTracks(onlyForYouTracks, {
        artist: {
            ban_genres: ['rap', 'r&b', 'metal', 'anime', 'soul', 'blues', 'punk'],
        },
    });
    
    // Подставьте id созданного плейлиста, после первого запуска кода, удалите комментарий
    Playlist.saveWithReplace({
        // id: 'вашеId',
        name: 'Незнакомый сет',
        tracks: Selector.sliceRandom(onlyForYouTracks, 60),
        description: Playlist.getDescription(onlyForYouTracks),
        randomCover: 'update',
    });
}
```
То здесь, в самом начале, есть объявленная переменная `let banTracks = [];` которая используется в 2-х местах в коде, но не содержит в себе никаких данных. Так как я переделал переменную **banTracks** на использование Бан-файла, а объявить еще одну переменную с таким же названием я не могу, то можно же спокойно удалять `let banTracks = [];` ? Она же не обязательная для работы скрипта?

**Reply by Chimildic** [reply]
В исходном варианте _banTracks_ не пустой массив. Он содержит объединение _savedTracks_ и _recentTracks_. 
Можете добавить треки из вашего кэша:
```js
 Combiner.push(banTracks, savedTracks, recentTracks, Cache.read('BannedTracks.json'));
```

**Reply by ScreNN** [reply]
То есть, поменять код на такой? Вернуть `let banTracks = [];` а `let banTracks = Cache.read('BannedTracks.json');` удалить?
```
// Плейлист дня из Яндекса
function getYandexDailyPlaylist(){

    let banTracks = [];

    let savedTracks = Source.getSavedTracks();
    let playlistTracks = gooex.Playlist.getTracks('109981031', 'yamusic-daily');
    let strTracks = yandexTracksToString_(playlistTracks);
    let spotifyTracks = Search.multisearchTracks(strTracks);

    Combiner.push(banTracks, savedTracks, Cache.read('BannedTracks.json'));
```

**Reply by Chimildic** [reply]
История прослушиваний _recentTracks_ тоже желательна, но смотрите сами.

**Reply by ScreNN** [reply]
То есть и в моем варианте все будет работать, но можно переделать как Вы сказали, и будет в принципе тоже самое? Оставлю тогда пока свой вариант, чтобы не путаться, и возможно добавлю историю прослушивания.

На счет истории прослушивания я думал. Этот лист тянется из "Плейлист дня" с Яндекса. Наверное, стоит все таки добавить историю прослушивания.

**Reply by Chimildic** [reply]
Вам быстрее скрипт запустить, чем ждать ответа)

**Reply by ScreNN** [reply]
Ладно, спасибо. **Bedmit** мне там тоже приготовил объяснение, пойду его донимать :)
